rubric_data/views.py

```diff
@@ -63,15 +60,22 @@ def denormalize(data):
         return []
     
     assignments_lookup = { assignment['id'] : assignment for assignment in data['assignments']}
-    criteria_lookup = { criterion['id'] : criterion 
+    # criterion IDs are NOT unique. Cat them with assignment IDs to create a unique key
+    criteria_lookup = { f"{assignment['id']}{criterion['id']}" : criterion
```

Comment:
Yeah, it appears that criterion IDs are only unique within the context of a rubric. From what I can tell, they are always prepended with an underscore, like `_368`. So that makes me think that they were designed to be concatenated with other values to create a unique key. Rating IDs are even more bizarre, as I've found the values `blank` and `blank_2` to be legit keys that are used. But that hasn't been a roadblock yet.

I am under the impression that an assignment can have one rubric, but a rubric can be used in many assignments. So in that way, I think it's better to stay focused on assignment and criteria. I could be wrong about that, and I think it points out one of the major challenges of the project: we don't yet have a clear data model that we can reference. Unless we can reference the canvas API docs?

Reply:
That's my impression too, that each course can have a pool of rubrics that can be reused across multiple assignments, but each assignment can have at most one Rubric associated with it. The canvas API docs aren't that great at making the relationships clear. You might have more luck looking directly at the models in the [instructure/canvas-lms](https://github.com/instructure/canvas-lms/tree/master/app/models) source repo (e.g. [rubric.rb](https://github.com/instructure/canvas-lms/blob/master/app/models/rubric.rb), [rubric_assessment.rb](https://github.com/instructure/canvas-lms/blob/master/app/models/rubric_assessment.rb), and [rubric_association.rb](https://github.com/instructure/canvas-lms/blob/master/app/models/rubric_association.rb)). From an end-user perspective,  I found [this](https://canvas.unl.edu/courses/38129/pages/using-rubrics) to be a good primer. 

As we talked about in dojo today, I think clarifying our understanding of the Canvas data model through diagrams and associated docs and then mapping that to our own explicit domain model in the code might help us moving forward.